test/test_nestedtensor.py

```diff
@@ -355,6 +355,54 @@ def test_device_checks(self, device):
         is_cuda = 'cuda' in str(device)
         self.assertEqual(nt.is_cuda, is_cuda)
 
+    @dtypes(torch.float, torch.float16, torch.double)
+    def test_nested_tensor_indexing(self, device, dtype):
+        # edge case: empty nested tensor
+        nt0 = torch.nested_tensor([])
+        self.assertRaisesRegex(
+            RuntimeError,
+            "cannot index an empty nested tensor",
+            lambda: nt0[0]
+        )
+        # normal case
+        x0 = torch.randn((2, 5), device=device)
+        x1 = torch.randn((3, 4), device=device)
+        nt = torch.nested_tensor([x0, x1])
+        # single index: only support integer in the batch dimension
+        self.assertEqual(nt[0], x0)
+        self.assertEqual(nt[-1], x1)
+        self.assertRaisesRegex(
+            IndexError,
+            "index 2 is out of bounds for nested tensor dimension 0 with size 2",
+            lambda: nt[2]
+        )
+        self.assertRaisesRegex(
+            IndexError,
+            "index -3 is out of bounds for nested tensor dimension 0 with size 2",
+            lambda: nt[-3]
+        )
+        self.assertRaises(NotImplementedError, lambda: nt[:])
+        self.assertRaises(NotImplementedError, lambda: nt[None])
+        self.assertRaises(NotImplementedError, lambda: nt[...])
+        # tuple of indices: only support integer in the batch dimension
+        #                 + all possible indexing in the original tensor dimensions
+        self.assertEqual(nt[0, 0, 0], x0[0, 0])
+        self.assertEqual(nt[0, 1, :], x0[1, :])
+        self.assertEqual(nt[1, ...], x1)
+        self.assertRaisesRegex(
+            IndexError,
+            "index 4 is out of bounds for dimension 1 with size 3",
+            lambda: nt[1, 4, 2]
+        )
+        self.assertRaises(NotImplementedError, lambda: nt[:, 1, 1])
+        # make sure indexing returns a view
```

Comment:
@albanD @drisspg - this will become more interesting once NestedTensor has autograd support.

Reply:
I considered nested tensor slicing, but the best way I could come up with is to only support slice 1: the sizes of the original tensors differ, so the stride can be irregular. Decided to leave it until we reach a better solution

Reply:
It is possible to support slicing and return a NestedTensor. It's a worthwhile extension, but I don't think we need it in the context of the PR.